### PR TITLE
typeclass extension

### PR DESCRIPTION
**Pull request by yiyunliu**
Passing all the [vrdt](https://github.com/jprider63/vrdt)  and [base](https://github.com/plum-umd/liquid-base) instance tests for the typeclass paper. 
Two remaining issues:
- --typeclass flag does not work in plugin mode. I used ghc interactive context to define functions like ==> and ==, but the interactive context is not available in TcM (monad where the plugin runs). We need to figure out a way that adds those extra definitions to the typing environment directly.
- some of the non-typeclass tests are still failing. Supposedly, when --typeclass is off, the behavior should be the same as the existing code and therefore all tests should pass. I probably messed up with the flagging somewhere.

and some general cleanup & additional comments to make the code maintainable.

### CODE REVIEW

File: src/Language/Haskell/Liquid/Types/RefType.hs [hunk 1]
```diff
@@ -1671,7 +1675,7 @@ grabArgs τs τ
   = reverse (τ:τs)
 
 isNonValueTy :: Type -> Bool
-isNonValueTy = GM.isPredType
+isNonValueTy = GM.isEmbeddedDictType -- GM.isPredType
```
Comment (nikivazou) [inline src/Language/Haskell/Liquid/Types/RefType.hs]: This looks suspicious

File: src/Language/Haskell/Liquid/Bare/Check.hs [hunk 2]
```diff
@@ -600,7 +613,7 @@ checkMBody γ emb _ sort (Def m c _ bs body) = checkMBody' emb sort γ' sp body
   where
     sp    = F.srcSpan m
     γ'    = L.foldl' (\γ (x, t) -> F.insertSEnv x t γ) γ xts
-    xts   = zip (fst <$> bs) $ rTypeSortedReft emb . subsTyVars_meet su <$> filter (not . isClassType) (ty_args trep)
+    xts   = zip (fst <$> bs) $ rTypeSortedReft emb . subsTyVars_meet su <$>  (ty_args trep)
```
Comment (nikivazou) [inline src/Language/Haskell/Liquid/Bare/Check.hs]: This looks suspicious

File: src/Language/Haskell/Liquid/Constraint/Generate.hs [hunk 3]
```diff
@@ -1043,10 +1047,11 @@ castTy γ t e _
 
 
 castTy' γ τ (Var x)
-  = do t <- trueTy τ
+  = do t <- trueTy (typeclass (getConfig γ)) τ
        -- tx <- varRefType γ x -- NV HERE: the refinements of the var x do not get into the 
        --                      -- environment. Check 
-       let ce = eCoerc (typeSort (emb γ) $ Ghc.expandTypeSynonyms $ varType x) 
+       let ce = if (noADT (getConfig γ)) then F.expr x
```
Comment (nikivazou) [inline src/Language/Haskell/Liquid/Constraint/Generate.hs]: This looks suspicious

File: tests/todo/RefinedEquality.hs [hunk 4]
```diff
@@ -0,0 +1,119 @@
+{-@ LIQUID "--reflection" @-}
```
Comment (yiyunliu) [inline tests/todo/RefinedEquality.hs:1]: not sure why this file is here

File: src/Language/Haskell/Liquid/Types/Types.hs [hunk 5]
```diff
@@ -1522,8 +1560,10 @@ mapExprReft f = mapReft g
   where
     g (MkUReft (F.Reft (x, e)) p) = MkUReft (F.Reft (x, f x e)) p
 
+-- const False (not dropping dict) is probably fine since there will not be refinement on
+-- dictionaries
 isTrivial :: (F.Reftable r, TyConable c) => RType c tv r -> Bool
-isTrivial t = foldReft False (\_ r b -> F.isTauto r && b) True t
+isTrivial t = foldReft (const False) False (\_ r b -> F.isTauto r && b) True t
```
Comment (yiyunliu) [inline src/Language/Haskell/Liquid/Types/Types.hs]: `isTrivial` should be flagged. `const False` means foldReft will not erase anything. Changing `const False` to `isClass` might fix some tests

File: src/Language/Haskell/Liquid/Types/RefType.hs [hunk 6]
```diff
@@ -1671,7 +1675,7 @@ grabArgs τs τ
   = reverse (τ:τs)
 
 isNonValueTy :: Type -> Bool
-isNonValueTy = GM.isPredType
+isNonValueTy = GM.isPredType-- GM.isEmbeddedDictType
```
Comment (yiyunliu) [inline src/Language/Haskell/Liquid/Types/RefType.hs:1699]: need to properly flag this once all tests pass

File: src/Language/Haskell/Liquid/Types/PredType.hs [hunk 7]
```diff
@@ -387,7 +391,7 @@ substRCon msg (_, RProp ss t1@(RApp c1 ts1 rs1 r1)) t2@(RApp c2 ts2 rs2 _) πs r
 
     su = F.mkSubst $ zipWith (\s1 s2 -> (s1, F.EVar s2)) (rvs t1) (rvs t2)
 
-    rvs      = foldReft False (\_ r acc -> rvReft r : acc) []
+    rvs      = foldReft (const False) False (\_ r acc -> rvReft r : acc) []
```
Comment (yiyunliu) [inline src/Language/Haskell/Liquid/Types/PredType.hs]: `const False` -> `isClass`

File: src/Language/Haskell/Liquid/Types/RefType.hs [hunk 8]
```diff
@@ -917,11 +917,15 @@ generalize :: (Eq tv, Monoid r) => RType c tv r -> RType c tv r
 generalize t = mkUnivs (zip (freeTyVars t) (repeat mempty)) [] t
 
 allTyVars :: (Ord tv) => RType c tv r -> [tv]
-allTyVars t = sortNub . fmap ty_var_value $ vs ++ vs'
+allTyVars = sortNub . allTyVars'
```
Comment (yiyunliu) [inline src/Language/Haskell/Liquid/Types/RefType.hs:932]: not sure about this

File: src/Language/Haskell/Liquid/Types/Fresh.hs [hunk 9]
```diff
@@ -268,20 +268,20 @@ refreshPs = mapPropM go
...
   | noHoles t = return (Nothing, x, t)
   | otherwise = (Just $ F.symbol x,x,) <$> mapReftM tx t
   where
-    tx r | hasHole r = refresh r
+    tx r | hasHole r = refresh allowTC r
          | otherwise = return r
 
 noHoles :: (F.Reftable r, TyConable c) => RType c tv r -> Bool
-noHoles = and . foldReft False (\_ r bs -> not (hasHole r) : bs) []
+noHoles = and . foldReft (const False) False (\_ r bs -> not (hasHole r) : bs) []
```
Comment (yiyunliu) [inline src/Language/Haskell/Liquid/Types/Fresh.hs]: flag

File: src/Language/Haskell/Liquid/Transforms/CoreToLogic.hs [hunk 10]
```diff
@@ -357,37 +363,41 @@ dataConProj dm de d x i = [(symbol x, t), (GM.simplesymbol x, t)]
...
+       e2 <- coreToLg allowTC etrue
        return $ EIte p e2 e1
 
-toPredApp :: C.CoreExpr -> LogicM Expr
-toPredApp p = go . Misc.mapFst opSym . splitArgs $ p
+toPredApp :: Bool -> C.CoreExpr -> LogicM Expr
+toPredApp allowTC p = go . Misc.mapFst opSym . splitArgs allowTC $ p
   where
-    opSym = fmap GM.dropModuleNames . tomaybesymbol
+    opSym = fmap GM.dropModuleNamesAndUnique . tomaybesymbol
```
Comment (yiyunliu) [inline src/Language/Haskell/Liquid/Transforms/CoreToLogic.hs:381]: should I flag this?

File: src/Language/Haskell/Liquid/Transforms/CoreToLogic.hs [hunk 11]
```diff
@@ -240,48 +243,51 @@ isMeasureArg x
...
 
-coreToLg :: C.CoreExpr -> LogicM Expr
-coreToLg (C.Let b e)
-  = subst1 <$> coreToLg e <*>  makesub b
-coreToLg (C.Tick _ e)          = coreToLg e
-coreToLg (C.App (C.Var v) e)
-  | ignoreVar v                = coreToLg e
-coreToLg (C.Var x)
+coreToLg :: Bool -> C.CoreExpr -> LogicM Expr
+coreToLg allowTC  (C.Let (C.NonRec x (C.Coercion c)) e)
```
Comment (yiyunliu) [inline src/Language/Haskell/Liquid/Transforms/CoreToLogic.hs:259]: I believe this is added new case is an improvement regardless of whether the typeclass flag is on. Might turn out to be problematic

File: src/Language/Haskell/Liquid/Transforms/CoreToLogic.hs [hunk 12]
```diff
@@ -55,25 +56,27 @@ import           Language.Haskell.Liquid.Types.RefType
...
-inlineSpecType v = fromRTypeRep $ rep {ty_res = res `strengthen` r , ty_binds = xs}
+inlineSpecType :: Bool -> Var -> SpecType
+inlineSpecType  allowTC v = fromRTypeRep $ rep {ty_res = res `strengthen` r , ty_binds = xs}
   where
     r              = MkUReft (mkR (mkEApp f (mkA <$> vxs))) mempty
     rep            = toRTypeRep t
     res            = ty_res rep
     xs             = intSymbol (symbol ("x" :: String)) <$> [1..length $ ty_binds rep]
-    vxs            = dropWhile (isClassType . snd) $ zip xs (ty_args rep)
+    vxs            = dropWhile (isErasable . snd) $ zip xs (ty_args rep)
```
Comment (yiyunliu) [inline src/Language/Haskell/Liquid/Transforms/CoreToLogic.hs:72]: flag

File: src/Language/Haskell/Liquid/Transforms/CoreToLogic.hs [hunk 13]
```diff
@@ -55,25 +56,27 @@ import           Language.Haskell.Liquid.Types.RefType
...
 
-logicType :: (Reftable r) => Type -> RRType r
-logicType τ      = fromRTypeRep $ t { ty_binds = bs, ty_args = as, ty_refts = rs}
+logicType :: (Reftable r) => Bool -> Type -> RRType r
+logicType allowTC τ      = fromRTypeRep $ t { ty_binds = bs, ty_args = as, ty_refts = rs}
   where
     t            = toRTypeRep $ ofType τ
-    (bs, as, rs) = unzip3 $ dropWhile (isClassType . Misc.snd3) $ zip3 (ty_binds t) (ty_args t) (ty_refts t)
+    (bs, as, rs) = unzip3 $ dropWhile (isErasable . Misc.snd3) $ zip3 (ty_binds t) (ty_args t) (ty_refts t)
+    isErasable   = if allowTC then isEmbeddedClass else isClassType
```
Comment (yiyunliu) [inline src/Language/Haskell/Liquid/Transforms/CoreToLogic.hs:58]: flag

File: src/Language/Haskell/Liquid/GHC/Plugin.hs [hunk 14]
```diff
@@ -426,8 +426,9 @@ processModule LiquidHaskellContext{..} = do
   -- Due to the fact the internals can throw exceptions from pure code at any point, we need to
   -- call 'evaluate' to force any exception and catch it, if we can.
 
+  -- what to do when maketargetspec is already in the monad?
   result <-
-    (liftIO $ evaluate (makeTargetSpec moduleCfg lhModuleLogicMap targetSrc bareSpec dependencies))
+    makeTargetSpec moduleCfg lhModuleLogicMap targetSrc bareSpec dependencies
```
Comment (yiyunliu) [inline src/Language/Haskell/Liquid/GHC/Plugin.hs]: not using evaluate could be a potential problem

File: src/Language/Haskell/Liquid/GHC/API.hs [hunk 15]
```diff
@@ -46,7 +46,8 @@ module Language.Haskell.Liquid.GHC.API (
 #endif
 #endif
 
-  ) where
+  , gHC_REAL
```
Comment (nikivazou) [inline src/Language/Haskell/Liquid/GHC/API.hs]: This breaks the build

**Comment by nikivazou** [comment]
All the tests seem to pass!!!  
We need 
- [ ] an "easy merging step" 
- [ ] add documentation in https://github.com/ucsd-progsys/liquidhaskell/blob/develop/docs/mkDocs/docs/specifications.md#modules-with-code-type-classes
- [ ] add the `typeclass` option extension (https://github.com/ucsd-progsys/liquidhaskell/blob/develop/docs/mkDocs/docs/options.md) 
- [ ] add a `.csv` time comparison (as in https://github.com/ucsd-progsys/liquidhaskell/blob/develop/tests/logs/compare-PLE.csv) 

Let's schedule a merging call on Monday, July 12. @ranjitjhala do you want to join?

**Comment by nikivazou** [comment]
@yiyunliu what happens if we try to run the typeclasses with the ghc polugin? Concretely, 
1) what is the exact problem why these two do not work together? 
2) what is the error you get if you try to use them together? (can we make it throw an explanatory error message)

**Comment by yiyunliu** [comment]
> @yiyunliu what happens if we try to run the typeclasses with the ghc polugin? Concretely,
> 
>     1. what is the exact problem why these two do not work together?
> 
>     2. what is the error you get if you try to use them together? (can we make it throw an explanatory error message)

The problem with running typeclasses with the ghc plugin has been fixed. These two didn't work together in the past because we were having trouble overwriting the definition of (==) so its argument type doesn't have to be an instance of Eq. @jprider63 found out a way to do that in the plugin mode as well.

**Comment by ranjitjhala** [comment]
Super news re plug-in! Any idea why CI is sad?

**Comment by jprider63** [comment]
I believe we fixed the CI issues in #1844, so CI should work after merging that.

**Comment by ranjitjhala** [comment]
And do you know what the deal is with the conflicts?

**Comment by yiyunliu** [comment]
I resolved the conflicts, but there's some issue with haddock which is why everything except cabal_810 is failing. I'll fix that.

**Comment by jprider63** [comment]
Seems like everything except the dependency resolution for cabal 9 passed!

**Review by nikivazou — APPROVED** [review]

**Comment by nikivazou** [comment]
Looks great! Let's merge if circle accepts!

**Comment by ranjitjhala** [comment]
OMG green light! Do you want to merge? I can deal with any conflicts in my PR?

**Comment by yiyunliu** [comment]
@nikivazou @jprider63 and I discussed today and we decided to merge once the tests all pass.
Also, let me know if you have any issues merging your PR due to the type class changes.

**Comment by ranjitjhala** [comment]
OMG thank you!!!

**Comment by ranjitjhala** [comment]
(Eagerly awaiting blog post -- or point me to some text/draft if you want me to make a pass @jprider63  and @yiyunliu )

**Comment by jprider63** [comment]
Whoohoo! I'm glad we've finally merged this! I've started a draft blog post. I'll share it on slack.

**Pull request merged**